File: apps/whispering/src/lib/operations/media.ts
import { tauri } from '#platform/tauri';
import { goto } from '$app/navigation';
import { log, report } from '$lib/report';
import { deviceConfig } from '$lib/state/device-config.svelte';
import { settings } from '$lib/state/settings.svelte';

// The one best-effort side effect for recording: pause whatever the system is
// playing while recording, resume it after. Recording never waits on this and
// never fails because of it.
//
// `chain` is the entire state: a promise resolving to the opaque session tokens
// we currently have paused (`[]` when nothing is paused). Every pause/resume
// tacks itself onto the tail, so the backend calls run strictly one after
// another: a late resume can never race a fresh pause from a quick
// stop-then-restart. The resolved value answers the only question resume needs
// ("which sessions did I pause?") and doubles as the "currently paused" flag.
// Both helpers always resolve, so the chain never wedges.
//
// Tokens are opaque platform identities (macOS output-active bundle ids /
// Windows AUMID / Linux MPRIS bus name); the frontend only ever round-trips them
// back to the backend.

let chain: Promise<string[]> = Promise.resolve([]);

function shouldPausePlayback(): boolean {
	return Boolean(tauri && settings.get('recording.pausePlayback'));
}

async function pausePlayingSessions(): Promise<string[]> {
	if (!tauri) return [];
	// `pause()` is infallible across IPC: Rust logs any platform failure and
	// reports "paused nothing". The try/catch only guards an unexpected invoke
	// rejection (e.g. the command going missing), never a playback error.
	try {
		return await tauri.media.pause();
	} catch (error) {
		log.warn(new Error(`Failed to pause playback: ${String(error)}`));
		return [];
	}
}

async function resumeSessions(sessions: string[]): Promise<void> {
	if (!tauri || sessions.length === 0) return;
	// `resume()` is infallible across IPC, mirroring `pause()`.
	try {
		await tauri.media.resume(sessions);
	} catch (error) {
		log.warn(new Error(`Failed to resume playback: ${String(error)}`));
	}
}

// The feature is on by default, so the first time it actually pauses something
// we explain it once (per device): on-by-default should be discoverable and
// consensual, not a silent surprise. Fires only when a real session was paused,
// never on a no-op pause.
function explainFirstPauseOnce(): void {
	if (deviceConfig.get('notices.pausePlaybackExplained')) return;
	deviceConfig.set('notices.pausePlaybackExplained', true);
	report.info({
		title: 'Paused your playback while recording',
		description:
			'Whispering pauses media while it captures your voice, then resumes it. You can turn this off anytime.',
		action: {
			label: 'Recording settings',
			onClick: () => goto('/settings/recording'),
		},
	});
}

export const recordingMedia = {
	/** Pause active playback if enabled. Fire-and-forget: recording never waits. */
	pause(): void {
		if (!shouldPausePlayback()) return;
		// Already paused? Keep that set; otherwise pause what's playing now.
		chain = chain.then(async (paused) => {
			const next = paused.length > 0 ? paused : await pausePlayingSessions();
			if (next.length > 0) explainFirstPauseOnce();
			return next;
		});
	},

	/**
	 * Resume whatever the matching `pause()` paused. A no-op when nothing was
	 * paused, so every stop/cancel/start-failure path can call it blindly.
	 */
	resume(): void {
		chain = chain.then(async (paused) => {
			await resumeSessions(paused);
			return [];
		});
	},
};

File: apps/whispering/src-tauri/src/media.rs
//! Pause and resume system media playback around recording.
//!
//! The frontend `recordingMedia` chain calls `pause_playback` on capture start
//! and `resume_playback` on capture end, trading opaque `String` session tokens:
//! `pause_playback` returns a token for each session it paused, and the frontend
//! hands that same set back to `resume_playback`. Token contents are
//! platform-private (macOS output-active bundle ids, Windows AUMIDs, Linux MPRIS
//! bus names) and the frontend never interprets them.
//!
//! Recording never waits on, and never fails because of, playback control: both
//! commands are infallible across IPC and the caller fires and forgets. Platform
//! failures are logged on the Rust side, never surfaced to the frontend (which
//! has no recovery for them). One platform module is compiled per target;
//! unsupported targets are silent no-ops.

#[cfg(target_os = "linux")]
mod linux;
#[cfg(target_os = "macos")]
mod macos;
#[cfg(target_os = "windows")]
mod windows;

/// Pause every system media session currently playing. Returns one opaque token
/// per session paused, to hand back to `resume_playback`. Pausing is gated to
/// never *start* playback: only sessions observed playing are touched, and the
/// dedicated pause command (never a play/pause toggle) is sent.
#[tauri::command]
#[specta::specta]
pub async fn pause_playback() -> Vec<String> {
    // Infallible across IPC: recording never waits on, nor fails because of,
    // playback control, and the frontend has no recovery for a pause failure.
    // A platform failure is logged here (the cause belongs in the Rust log, not
    // the wire) and reported as "paused nothing", so the resume side has no
    // token to act on.
    #[cfg(target_os = "macos")]
    let result = macos::pause_playing().await;
    #[cfg(target_os = "windows")]
    let result = windows::pause_playing().await;
    #[cfg(target_os = "linux")]
    let result = linux::pause_playing().await;
    #[cfg(not(any(target_os = "macos", target_os = "windows", target_os = "linux")))]
    let result: Result<Vec<String>, String> = Ok(Vec::new());

    result.unwrap_or_else(|err| {
        log::warn!("pause_playback failed, leaving system playback untouched: {err}");
        Vec::new()
    })
}

/// Resume the sessions named by `sessions`, which must be tokens returned by a
/// prior `pause_playback`. A session that vanished, was already resumed by the
/// user, or can't be resumed is silently skipped. Safety rule: we only ever send
/// *play* to a session we personally paused.
#[tauri::command]
#[specta::specta]
pub async fn resume_playback(sessions: Vec<String>) {
    // Infallible across IPC, mirroring `pause_playback`: a resume failure is not
    // something the frontend can act on, so the cause is logged here rather than
    // returned.
    #[cfg(target_os = "macos")]
    let result = macos::resume(sessions).await;
    #[cfg(target_os = "windows")]
    let result = windows::resume(sessions).await;
    #[cfg(target_os = "linux")]
    let result = linux::resume(sessions).await;
    #[cfg(not(any(target_os = "macos", target_os = "windows", target_os = "linux")))]
    let result: Result<(), String> = {
        let _ = sessions;
        Ok(())
    };

    if let Err(err) = result {
        log::warn!("resume_playback failed: {err}");
    }
}
